lib/cnet/eth/eth_tx.c

```diff
@@ -17,17 +17,44 @@
 #include <cnet_node_names.h>
 #include "eth_tx_priv.h"        // for eth_tx_node_ctx_t, ETH_TX_NEXT_MAX
 #include "chnl_callback_priv.h"
+#include "cnet_pcb.h"
 
 static struct eth_tx_node_main eth_tx_main;
 
+static inline int
+eth_tx_node_pktmbuf_has_userdata(pktmbuf_t *mbuf)
+{
+    struct pcb_entry *pcb = mbuf->userptr;
+    if (unlikely(!pcb))
+        return -1;
+
+    if (pcb->ip_proto == IPPROTO_UDP)
```

Comment:
If we remove these two lines and at the bottom of the routine we return ETH_TX_NEXT_PKT_CALLBACK. This will reduce the number of 'if' tests and if a packet is not TCP with data the packet will be sent to the callback.

Does that make sense to only look at TCP packets and everything else can be sent to the callback?
Still thinking about this change as it is trying to optimize for a traffic generator use case.

Reply:
What about other protocols e.g. ICMPv6? 
Do you think we should report control protocols to the user application?
My idea is we do not need to report control protocols because they should be handled by the stack itself, just like in the linux kernel: the user space only gets user data (regardless of zero size user packets like traffic generators)